File: app/javascript/controllers/ceps_controller.js
import { Controller } from "stimulus";
import { post } from "@rails/request.js";

export default class extends Controller {
    static targets = ["cep", "street", "city", "state", "form", "feedback"];

    reset() {
        this.formTarget.reset();
    }

    async fetchAddress() {
        if (this.cepTarget.value.length >= 8) {
            const response = await post("/fetch_address", {
                responseKind: "json",
                body: { cep: this.cepTarget.value },
            });
            if (response.ok) {
                const data = await response.json;
                if (data.address === "formato inválido") {
                    this.streetTarget.value = "";
                    this.cityTarget.value = "";
                    this.stateTarget.value = "";
                    this.feedbackTarget.innerHTML =
                        "<p class='text-danger text-center'>formato inválido</p>";
                } else if (data.address === "CEP não encontrado") {
                    this.streetTarget.value = "";
                    this.cityTarget.value = "";
                    this.stateTarget.value = "";
                    this.feedbackTarget.innerHTML =
                        "<p class='text-danger text-center'>CEP não encontrado</p>";
                } else {
                    this.feedbackTarget.innerHTML = "";
                    this.streetTarget.value = data.address.street;
                    this.cityTarget.value = data.address.city;
                    this.stateTarget.value = data.address.state;
                }
            }
        }
    }
}

File: app/controllers/ceps_controller.rb
class CepsController < ApplicationController
  skip_before_action :authenticate_user!, only: %i[calculate_shipment]

  def fetch_address
    authorize :cep
    begin
      address = ViaCep::Address.new(params[:cep])
    rescue ViaCep::Errors::InvalidZipcodeFormat
      address = 'formato inválido'
    rescue ViaCep::Errors::ZipcodeNotFound
      address = 'CEP não encontrado'
    end
    respond_to do |format|
      format.json { render json: { address: address } }
    end
  end

  def calculate_shipment
    skip_authorization
    begin
      address = ViaCep::Address.new(params[:cep])
      destination = "#{address.street}, #{address.city}, #{address.state}"
      shipment_distance = Geocoder::Calculations.distance_between(CARREFOUR_COORDS, destination)
      shipment = shipment_distance < 100 ? 1499 : 1499 + shipment_distance.round
    rescue ViaCep::Errors::InvalidZipcodeFormat
      address = 'formato inválido'
    rescue ViaCep::Errors::ZipcodeNotFound
      address = 'CEP não encontrado'
    end
    respond_to do |format|
      format.json { render json: { address: address, shipment: shipment } }
    end
  end
end
